aten/src/ATen/Context.cpp

```diff
@@ -295,8 +295,24 @@ bool Context::hasLAPACK() {
 }
 
 at::QEngine Context::qEngine() const {
-  // If wasn't explicitly set - take the last one available
-  return quantized_engine.value_or(supportedQEngines().back());
+  static auto _quantized_engine = []() {
+    at::QEngine qengine = at::kNoQEngine;
+#ifdef USE_PYTORCH_QNNPACK
+    qengine = at::kQNNPACK;
+#endif // USE_PYTORCH_QNNPACK
+
+#if AT_MKLDNN_ENABLED()
+    qengine = at::kONEDNN;
+#endif
+
+#ifdef USE_FBGEMM
+    if (fbgemm::fbgemmSupportedCPU()) {
+      qengine = at::kFBGEMM;
+    }
+#endif
+    return qengine;
+  }();
```

Comment:
What's the purpose of using a lambda here? I feel it would be easier to read without it.

Reply:
Also, is it possible to have return statements inside `ifdef` blocks? If so that would be cleaner I think

Reply:
this is to make sure the function is only run once during first call I think, sure I can move around return

Reply:
how to return in `ifdef`?